File: src/main/java/me/itzisonn_/meazy/parser/ast/statement/BreakStatement.java
package me.itzisonn_.meazy.parser.ast.statement;

import me.itzisonn_.meazy.instruction.InstructionsSet;
import me.itzisonn_.meazy.parser.ast.ProgramUnit;
import me.itzisonn_.meazy.runtime.environment.Environment;
import me.itzisonn_.meazy.runtime.environment.EnvironmentUtils;
import me.itzisonn_.meazy.runtime.environment.LoopEnvironment;
import org.jspecify.annotations.NullMarked;

@NullMarked
public class BreakStatement implements LocalStatement {
    @Override
    public void emit(InstructionsSet instructions, Environment environment, ProgramUnit parent) {
        LoopEnvironment loopEnvironment = EnvironmentUtils.getParentOrSelf(environment, LoopEnvironment.class).orElseThrow(
                () -> new IllegalArgumentException("Parent environment for BREAK statement must be LoopEnvironment TODO")
        );

        instructions.gotoLabel(loopEnvironment.getEndLabel());
    }

    @Override
    public boolean alwaysReturns() {
        return false;
    }
}

File: src/main/kotlin/me/itzisonn_/meazy/parser/ast/statement/BreakStatement.kt
package me.itzisonn_.meazy.parser.ast.statement

import me.itzisonn_.meazy.instruction.InstructionsSet
import me.itzisonn_.meazy.parser.ast.ParentMap
import me.itzisonn_.meazy.parser.ast.ProgramUnit
import me.itzisonn_.meazy.runtime.environment.Environment
import me.itzisonn_.meazy.runtime.environment.LoopEnvironment
import me.itzisonn_.meazy.runtime.environment.getParentOrSelf

class BreakStatement : LocalStatement {
    override val children = setOf<ProgramUnit>()

    context(parents: ParentMap)
    override fun emit(instructions: InstructionsSet, environment: Environment) {
        val loopEnvironment = environment.getParentOrSelf<LoopEnvironment>()
            ?: error("Parent environment for BREAK statement must be LoopEnvironment TODO")

        instructions.gotoLabel(loopEnvironment.getEndLabel())
    }

    override fun alwaysReturns() = false
}
